sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala

```diff
@@ -181,3 +188,237 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+  extends ExpressionWithRandomSeed with BinaryLike[Expression] with Nondeterministic {
+  def this(length: Expression) = this(length, UnresolvedSeed)
+
+  override def nullable: Boolean = false
+  override def dataType: DataType = StringType
+  override def stateful: Boolean = true
+  override def left: Expression = length
+  override def right: Expression = seedExpression
+
+  /**
+   * Record ID within each partition. By being transient, the Random Number Generator is
+   * reset every time we serialize and deserialize and initialize it.
+   */
+  @transient protected var rng: XORShiftRandom = _
+
+  @transient protected lazy val seed: Long = seedExpression match {
+    case e if e.dataType == IntegerType => e.eval().asInstanceOf[Int]
+    case e if e.dataType == LongType => e.eval().asInstanceOf[Long]
+  }
+  override protected def initializeInternal(partitionIndex: Int): Unit = {
+    rng = new XORShiftRandom(seed + partitionIndex)
+  }
+
+  override def withNewSeed(newSeed: Long): Expression = RandStr(length, Literal(newSeed, LongType))
+  override def withNewChildrenInternal(newFirst: Expression, newSecond: Expression): Expression =
+    RandStr(newFirst, newSecond)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    var result: TypeCheckResult = TypeCheckResult.TypeCheckSuccess
+    def requiredType = "INT or SMALLINT"
+    Seq((length, "length", 0),
+      (seedExpression, "seedExpression", 1)).foreach {
+      case (expr: Expression, name: String, index: Int) =>
+        if (result == TypeCheckResult.TypeCheckSuccess) {
+          if (!expr.foldable) {
+            result = DataTypeMismatch(
+              errorSubClass = "NON_FOLDABLE_INPUT",
+              messageParameters = Map(
+                "inputName" -> name,
+                "inputType" -> requiredType,
+                "inputExpr" -> toSQLExpr(expr)))
+          } else expr.dataType match {
+            case _: ShortType | _: IntegerType =>
+            case _: LongType if index == 1 =>
+            case _ =>
+              result = DataTypeMismatch(
+                errorSubClass = "UNEXPECTED_INPUT_TYPE",
+                messageParameters = Map(
+                  "paramIndex" -> ordinalNumber(index),
+                  "requiredType" -> requiredType,
+                  "inputSql" -> toSQLExpr(expr),
+                  "inputType" -> toSQLType(expr.dataType)))
+          }
+        }
+    }
+    result
+  }
+
+  override def evalInternal(input: InternalRow): Any = {
+    val numChars: Int = length.eval(input) match {
+      case i: Int => i
+      case n: Long => n.toInt
+      case s: Short => s.toInt
+    }
+    val bytes = new Array[Byte](numChars)
+    (0 until numChars).foreach { i =>
+      val num = (rng.nextInt() % 30).abs
+      num match {
+        case _ if num < 10 =>
+          bytes.update(i, ('0' + num).toByte)
+        case _ if num < 20 =>
+          bytes.update(i, ('a' + num - 10).toByte)
```

Comment:
And how your algorithm returns 'z', for instance?
```
'a' = 97, 'z' = 122
122 = 97 + num - 10 -> num = 35
```
the first case:
```
'0' = 48
122 = 48 + num -> num = 74
```
the last one:
```
'A' = 65
122 = 65 + num - 20 -> num = 77
```

To get 'z', num must be greater than `(rng.nextInt() % 30).abs`.

Your expression description says:
```
characters are chosen uniformly at random from the following pool of characters: 0-9, a-z, A-Z
```

There is some contradiction, isn't it?

Reply:
You are correct, this was a bug in the algorithm. The random number range should actually be between 0 and 61, inclusive. Then the range [0, 9] corresponds to digits 0-9, [10-35] corresponds to characters a-z, and [36-61] corresponds to A-Z. I updated the algorithm to fix it and added a comment here.